### PR TITLE
OAK-4101 Consider separate external (group) principal management

### PR DESCRIPTION
**Pull request by tripodsan**
I created this pull request in order to be able to discuss the proposed patch in https://issues.apache.org/jira/browse/OAK-4101. I'm not the original author.

### CODE REVIEW

File: oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/basic/DefaultSyncContext.java [hunk 1]
```diff
@@ -400,10 +400,15 @@ protected Group createGroup(@Nonnull ExternalGroup externalGroup) throws Reposit
                 principal,
                 PathUtils.concatRelativePaths(config.group().getPathPrefix(), externalGroup.getIntermediatePath())
         );
-        group.setProperty(REP_EXTERNAL_ID, valueFactory.createValue(externalGroup.getExternalId().getString()));
+        setExternalId(group, externalGroup);
         return group;
     }
 
+    private void setExternalId(@Nonnull Authorizable authorizable, @Nonnull ExternalIdentity externalIdentity) throws RepositoryException {
```
Comment (tripodsan) [inline oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/basic/DefaultSyncContext.java:407]: minor: JavaDoc missing

File: oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        } else {
+            // retrieve membership of the given external user (up to the configured
+            // depth) and add (or replace) the rep:externalPrincipalNames property
+            // with the accurate collection of principal names.
+            try {
+                Value[] vs;
+                if (depth <= 0) {
+                    vs = new Value[0];
+                } else {
+                    Set<String> principalsNames = new HashSet();
```
Comment (tripodsan) [inline oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java:125]: better: `new HashSet<String>()`

File: oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+                    collectPrincipalNames(principalsNames, external.getDeclaredGroups(), depth);
+                    vs = createValues(principalsNames);
+                }
+                auth.setProperty(ExternalIdentityConstants.REP_EXTERNAL_PRINCIPAL_NAMES, vs);
+            } catch (ExternalIdentityException e) {
+                log.error("Failed to synchronize membership information for external identity " + external.getId(), e);
+                return;
```
Comment (tripodsan) [inline oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java:132]: not needed

File: oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+            ExternalIdentity extId = idp.getIdentity(ref);
+            if (extId instanceof ExternalGroup) {
+                principalNames.add(extId.getPrincipalName());
+                // recursively apply further membership until the configured depth is reached
+                if (depth > 1) {
+                    collectPrincipalNames(principalNames, extId.getDeclaredGroups(), depth - 1);
+                }
+            } else {
+                log.debug("Not an external group => ignore");
+                continue;
```
Comment (tripodsan) [inline oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java:149]: not needed

File: oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+                    return syncGroup((ExternalGroup) identity, group);
+                } else {
+                    // external group has never been synchronized before:
+                    // don't sync external groups into the repository internal user management
+                    // but limit synchronized information to group-principals stored
+                    // separately with each external user such that the subject gets
+                    // properly populated upon login
+                    ExternalIdentityRef ref = identity.getExternalId();
+
+                    log.debug("ExternalGroup " + ref.getString() + ": Not synchronized as authorizable Group into the repository.");
```
Comment (tripodsan) [inline oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java:93]: use: `{}` notation.

File: oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+                    vs = createValues(principalsNames);
+                }
+                auth.setProperty(ExternalIdentityConstants.REP_EXTERNAL_PRINCIPAL_NAMES, vs);
+            } catch (ExternalIdentityException e) {
+                log.error("Failed to synchronize membership information for external identity " + external.getId(), e);
+                return;
+            }
+        }
+    }
+
```
Comment (tripodsan) [inline oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java:136]: minor: Javadoc....

File: oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/ExternalGroupPrincipalProvider.java [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+        // no group principals defined
+        return ImmutableSet.of();
+    }
+
+    @CheckForNull
+    private Result findPrincipals(@Nonnull String queryString, boolean exactMatch) {
+        try {
+            Map<String, ? extends PropertyValue> bindings = Collections.singletonMap("princNames", PropertyValues.newString(queryString));
+            String op = (exactMatch) ? " = " : " LIKE ";
+            String statement = "SELECT '" + REP_EXTERNAL_PRINCIPAL_NAMES + "' FROM [rep:User] WHERE PROPERTY(["
```
Comment (tripodsan) [inline oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/ExternalGroupPrincipalProvider.java:189]: consider use of: `String.format()`. this is barely readable :-)

File: oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/ExternalGroupPrincipalProvider.java [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+        return ImmutableSet.of();
+    }
+
+    @Nonnull
+    @Override
+    public Set<? extends Principal> getPrincipals(@Nonnull String userID) {
+        try {
+            return getGroupPrincipals(userManager.getAuthorizable(userID));
+        } catch (RepositoryException e) {
+            log.debug(e.getMessage());
```
Comment (tripodsan) [inline oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/ExternalGroupPrincipalProvider.java:135]: not recommended to use the message as is here. better: `log.debug("{}", e.getMessage());`

File: oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/ExternalGroupPrincipalProvider.java [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+                    groupPrincipals.add(new ExternalGroupPrincipal(principalName));
+                }
+                return groupPrincipals;
+            }
+        }
+        // no group principals defined
+        return ImmutableSet.of();
+    }
+
+    @CheckForNull
```
Comment (tripodsan) [inline oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/ExternalGroupPrincipalProvider.java:184]: maybe some explanation of what is exactly searched for. this helps understanding the code and counter-check if the query formulates what is intended.

File: oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/basic/DefaultSyncContext.java (hunk 1 above)
Comment (anchela) [inline oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/basic/DefaultSyncContext.java:407]: done

File: oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java (hunk 2 above)
Comment (anchela) [inline oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java:125]: done

File: oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java (hunk 3 above)
Comment (anchela) [inline oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java:132]: done

File: oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java (hunk 4 above)
Comment (anchela) [inline oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java:149]: done

File: oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java (hunk 5 above)
Comment (anchela) [inline oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java:93]: done

File: oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java (hunk 6 above)
Comment (anchela) [inline oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/PrincipalSyncContext.java:136]: done

File: oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/ExternalGroupPrincipalProvider.java (hunk 8 above)
Comment (anchela) [inline oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/ExternalGroupPrincipalProvider.java:135]: done

File: oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/ExternalGroupPrincipalProvider.java (hunk 9 above)
Comment (anchela) [inline oak-auth-external/src/main/java/org/apache/jackrabbit/oak/spi/security/authentication/external/impl/principal/ExternalGroupPrincipalProvider.java:184]: done
see also test-cases that should verify that the query works.

**Comment by tripodsan** [comment]
closing due to inactivity.

**Pull request closed without merging**